qutebrowser/components/utils/blockutils.py

```diff
@@ -0,0 +1,170 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2020 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+
+"""Code that is shared between the host blocker and Brave ad blocker."""
+
+import typing
+import os
+import functools
+import threading
+
+from PyQt5.QtCore import QUrl
+
+from qutebrowser.api import downloads, message, config
+
+
+class FakeDownload(downloads.TempDownload):
+
+    """A download stub to use on_download_finished with local files."""
+
+    def __init__(self, fileobj: typing.IO[bytes]) -> None:
+        # pylint: disable=super-init-not-called
+        self.fileobj = fileobj
+        self.successful = True
+
+
+class BlocklistDownload:
+    """Download blocklists from the given URLs.
+
+    Attributes:
+        _urls: The URLs to download.
+        _user_cb_single:
+            A user-provided function to be called when a single download has
+            finished. The user is provided with the download object.
+        _user_cb_all:
+            A user-provided function to be called when all downloads have
+            finished. The first argument to the function is the number of
+            items downloaded.
+        _in_progress: The DownloadItems which are currently downloading.
+        _done_count: How many files have been read successfully.
+        _finished_registering_downloads:
+            Used to make sure that if all the downloads finish really quickly,
+            before all of the block-lists have been added to the download
+            queue, we don't call `_on_lists_downloaded`.
+        _started: Has the `initiate` method been called?
+        _finished: Has `_user_cb_all` been called?
+    """
+
+    def __init__(
+        self,
+        urls: typing.List[QUrl],
+        on_single_download: typing.Callable[[typing.IO[bytes]], typing.Any],
+        on_all_downloaded: typing.Callable[[int], typing.Any],
```

Comment:
I think it would be nicer to have them as [signals](https://doc.qt.io/qt-5/signalsandslots.html) which are typically used by Qt rather than callbacks.

To do so, you'd have to:

- Make `BlocklistDownload` inherit from `QObject`
- Add a `parent=None` argument as the last one
- Call `super().__init__(parent)`
- At class level, add something like:

```python
download_finished = pyqtSignal(io.IOBase)  # arg: the file object
all_downloads_finished = pyqtSignal(int)  # arg: download count
```

- Hook those signals up when creating `BlocklistDownload`, i.e. something like:

```python
dl.download_finished.connect(self._merge_file)
dl.all_downloads.finished.connect(self._on_lists_downloaded)
```

This will also make it a bit easier to use those (or wait for them) during the tests.